### ISSUE TITLE
"sort" causes Zed lake service crash

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#5236]
## tl;dr

A `zed serve` hosting a pool of the [zed-sample-data](https://github.com/brimdata/zed-sample-data/tree/main/zng) can be crashed by a `sort` query.

## Details

Repro is with Zed commit 3f5af44. This issue was spotted by our own @jameskerr while doing Zui development/testing.

First start a Zed lake service. I've not been able to repro this problem with a filesystem-based lake.

```
$ zed -version
Version: v1.17.0-31-g3f5af44b

$ rm -rf lake && zed -lake lake serve
{"level":"info","ts":1724174105.489831,"logger":"core","msg":"Started","auth_enabled":false,"root":"file:///Users/phil/lake","version":"v1.17.0-31-g3f5af44b"}
{"level":"info","ts":1724174105.49012,"logger":"httpd","msg":"Listening","addr":"[::]:9867"}
```

Then load the [ZNG format zed-sample-data](https://github.com/brimdata/zed-sample-data/tree/main/zng) and issue a `sort` query.

```
$ zed create -use zng
pool created: zng 2kvptA65o0HdU7tVRRpt1DhKbnt
Switched to branch "main" on pool "zng"

$ zed load *.zng.gz
(35/35) 51.49MB/51.49MB 0B/s 100.00%
2kvpuHy8843DfPgH6blJMNUBIcz committed

$ zed query 'sort'
Post "http://localhost:9867/query?ctrl=T": EOF
```

At this point the `zed serve` crashes with the following dump:

```
panic: runtime error: slice bounds out of range [:374700416] with capacity 655418

goroutine 226 [running]:
github.com/brimdata/zed.(*Arena).bytes_(0x140aa50?, 0x25ef560?)
	/Users/phil/work/zed/arena.go:191 +0x3dc
github.com/brimdata/zed.Value.Bytes({0x100000c00073b5f0, 0x4000001e00000000})
	/Users/phil/work/zed/value.go:232 +0x20a
github.com/brimdata/zed/runtime/sam/expr.(*DotExpr).Eval(0xc005ee6000, {0x2b44f58?, 0xc00657a000?}, {0x26a8a20?, 0x1?})
	/Users/phil/work/zed/runtime/sam/expr/dot.go:50 +0x21c
github.com/brimdata/zed/runtime/sam/expr.(*missingAsNull).Eval(0x3a65180?, {0x2b44f58?, 0xc00657a000?}, {0x100000c00073b5f0, 0x4000001e00000000})
	/Users/phil/work/zed/runtime/sam/expr/sort.go:137 +0x2f
github.com/brimdata/zed/runtime/sam/expr.(*Comparator).sortStableIndices(0xc005ee8000, {0xc015ff6000, 0x118200, 0x152c00})
	/Users/phil/work/zed/runtime/sam/expr/sort.go:33 +0x243
github.com/brimdata/zed/runtime/sam/expr.(*Comparator).SortStableReader(...)
	/Users/phil/work/zed/runtime/sam/expr/sort.go:250
github.com/brimdata/zed/runtime/sam/op/spill.(*MergeSort).Spill.func1()
	/Users/phil/work/zed/runtime/sam/op/spill/merge.go:67 +0x45
github.com/brimdata/zed/runtime/sam/op/spill.goWithContext.func1()
	/Users/phil/work/zed/runtime/sam/op/spill/merge.go:89 +0x23
created by github.com/brimdata/zed/runtime/sam/op/spill.goWithContext in goroutine 183
	/Users/phil/work/zed/runtime/sam/op/spill/merge.go:88 +0x8d
```

It looks like this problem has been with us since Zed commit ce9626e when the Arena changes landed via #5118.

**Comment by philrz** [comment]
A user in a [community Slack thread](https://brimdata.slack.com/archives/C03MW6XT7HC/p1724098664936219) reported a recent Zed service crash that sounded similar to this one. They were not able to share their repro data, but here was their stack dump and a high-level description of their repro steps:

> ![image](https://github.com/user-attachments/assets/35d745d2-8e90-40b9-b6e3-75b36585cded)
>
> So the query is :
>```
>"Search term"
>| ts:=time(ts) | sort ts
>| "Search term 2"
>```
>
> If I comment out `| sort ts` then it's all gud
> Sorting last (which I know is probably the more polite thing to do) works great

The fact that sorting last might trim their data set to avoid spill-to-disk gives me high confidence it's the same problem.

**Comment by philrz** [comment]
Verified in Zed commit 1ffe14f.

Repeating the repro steps, the `sort` now runs without crashing.

```
$ zed -version
Version: v1.17.0-48-g1ffe14f6

$ zed create -use zng
pool created: zng 2lf9ET4XSCpvUjJ3GOy3w4uxtMv
Switched to branch "main" on pool "zng"

$ zed load *.zng.gz
(35/35) 51.49MB/51.49MB 0B/s 100.00%
2lf9FIu6d3D9BYlO3hr95AnG1aJ committed

$ zed query 'sort'
{_path:"analyzer",ts:2018-03-24T17:34:05.233607Z,cause:"violation",analyzer_kind:"protocol",analyzer_name:"HTTP",uid:"CFiSQy37m2jbdIGzW",fuid:null(string),id:{orig_h:10.128.0.19,orig_p:60489(port=uint16),resp_h:10.47.26.55,resp_p:80(port)},failure_reason:"not a http reply line",failure_data:null(string)}
...
```

Thanks @mattnibs!

**Comment by philrz** [comment]
Since the fix in linked PR #5239 was in the Arena era and the Arena changes were reverted in #5278, I've re-verified that this issue has stayed gone. Repeating the repro steps now at current Zed commit 4a1ebf5:

```
$ zed -version
Version: v1.17.0-64-g4a1ebf5b

$ zed create -use zng
pool created: zng 2mAbfgvYvs1s2TOMyoL2kdbIGKh
Switched to branch "main" on pool "zng"

$ zed load *.zng.gz
(35/35) 51.49MB/51.49MB 0B/s 100.00%
2mAbgOhFoLAfj3a5R4RwWEOaKaJ committed

$ zed query 'sort'
{_path:"analyzer",ts:2018-03-24T17:34:05.233607Z,cause:"violation",analyzer_kind:"protocol",analyzer_name:"HTTP",uid:"CFiSQy37m2jbdIGzW",fuid:null(string),id:{orig_h:10.128.0.19,orig_p:60489(port=uint16),resp_h:10.47.26.55,resp_p:80(port)},failure_reason:"not a http reply line",failure_data:null(string)}
...
```
